ndcube/meta.py

```diff
@@ -0,0 +1,248 @@
+import copy
+import numbers
+import collections.abc
+
+import numpy as np
+
+__all__ = ["Meta"]
+
+
+class Meta(dict):
+    """
+    A sliceable object for storing metadata.
+
+    Metdata can be linked to a data axis which causes it to be sliced when the
+    standard Python numeric slicing API is applied to the object.
+    Specific pieces of metadata can be obtain using the dict-like str slicing API.
+    Metadata associated with an axis/axes must have the same length/shape as those axes.
+
+    Parameters
+    ----------
+    header: dict-like
+        The names and values of metadata.
+
+    comments: dict-like (optional)
+        Comments associated with any of the above pieces of metadata.
+
+    axes: dict-like (optional)
+        The axis/axes associated with the above metadata values.
+        Each axis value must be None (for no axis association), and `int`
+        or an iterable of `int` if the metadata is associated with multiple axes.
+        Metadata in header without a corresponding entry here are assumed to not
+        be associated with an axis.
+
+    data_shape: iterable of `int` (optional)
+        The shape of the data with which this metadata is associated.
+        Must be set if axes input is set.
+    """
+    def __init__(self, header=None, comments=None, axes=None, data_shape=None):
+        self.original_header = header
+
+        # Sanitize metadata values and instantiate class.
+        if header is None:
+            header = {}
```

Comment:
Code cov is saying quite a few places aren't covered. Unsure if you want to address this in this PR or in a future one.

Reply:
Almost all those places are errors being raised and over 90% of the diff is being hit.  I think that's fine for this PR.

Reply:
Works for me.